scripts/ci/testing/ci_run_airflow_testing.sh

```diff
@@ -109,66 +66,265 @@ function prepare_tests_to_run() {
+    echo
+    echo "${COLOR_BLUE}Memory available for Docker${COLOR_RESET}"
+    echo
+    echo "    ${MEMORY_AVAILABLE_FOR_DOCKER}"
+    echo
+    export MEMORY_AVAILABLE_FOR_DOCKER
+}
 
-    INTEGRATIONS=()
-    export INTEGRATIONS
 
-    if [[ ${TEST_TYPE:=} == "Integration" ]]; then
-        export ENABLED_INTEGRATIONS="${AVAILABLE_INTEGRATIONS}"
-        export RUN_INTEGRATION_TESTS="${AVAILABLE_INTEGRATIONS}"
+function get_available_cpus_in_docker() {
+    CPUS_AVAILABLE_FOR_DOCKER=$(docker run --entrypoint /bin/bash \
+        "${AIRFLOW_CI_IMAGE}" -c \
+        'grep -cE "cpu[0-9]+" </proc/stat')
+    echo
+    echo "${COLOR_BLUE}CPUS available for Docker${COLOR_RESET}"
+    echo
+    echo "    ${CPUS_AVAILABLE_FOR_DOCKER}"
+    echo
+    export CPUS_AVAILABLE_FOR_DOCKER
+}
+
+function get_maximum_parallel_test_jobs() {
+    if [[ ${MAX_PARALLEL_TEST_JOBS=} != "" ]]; then
+        echo
+        echo "${COLOR_YELLOW}Maximum parallel test jobs forced vi MAX_PARALLEL_TEST_JOBS = ${MAX_PARALLEL_TEST_JOBS}${COLOR_RESET}"
+        echo
     else
-        export ENABLED_INTEGRATIONS=""
-        export RUN_INTEGRATION_TESTS=""
+        MAX_PARALLEL_TEST_JOBS=${CPUS_AVAILABLE_FOR_DOCKER}
+        echo
+        echo "${COLOR_YELLOW}Maximum parallel test jobs set to number of CPUs available for Docker = ${MAX_PARALLEL_TEST_JOBS}${COLOR_RESET}"
+        echo
     fi
+    export MAX_PARALLEL_TEST_JOBS
+}
+
+# Cleans up runner before test execution.
+#  * Kills all running docker containers
+#  * System prune to clean all the temporary/unnamed images and left-over volumes
+#  * Print information about available space and memory
+#  * Kills stale semaphore locks
+function cleanup_runner() {
+    start_end::group_start "Cleanup runner"
+    kill_all_running_docker_containers
+    system_prune_docker
+    print_available_space_in_host
+    print_available_memory_in_host
+    get_available_memory_in_docker
+    get_available_cpus_in_docker
+    get_maximum_parallel_test_jobs
+    parallel::kill_stale_semaphore_locks
+    start_end::group_end
+}
+
+# Starts test types in parallel
+# test_types_to_run - list of test types (it's not an array, it is space-separate list)
+# ${@} - additional arguments to pass to test execution
+function run_test_types_in_parallel() {
+    start_end::group_start "Monitoring tests: ${test_types_to_run}"
+    parallel::monitor_progress
+    mkdir -p "${PARALLEL_MONITORED_DIR}/${SEMAPHORE_NAME}"
+    for TEST_TYPE in ${test_types_to_run}
+    do
+        export TEST_TYPE
+        mkdir -p "${PARALLEL_MONITORED_DIR}/${SEMAPHORE_NAME}/${TEST_TYPE}"
+        mkdir -p "${PARALLEL_MONITORED_DIR}/${SEMAPHORE_NAME}/${TEST_TYPE}"
+        export JOB_LOG="${PARALLEL_MONITORED_DIR}/${SEMAPHORE_NAME}/${TEST_TYPE}/stdout"
+        export PARALLEL_JOB_STATUS="${PARALLEL_MONITORED_DIR}/${SEMAPHORE_NAME}/${TEST_TYPE}/status"
+        # shellcheck disable=SC2086
+        parallel --ungroup --bg --semaphore --semaphorename "${SEMAPHORE_NAME}" \
```

Comment:
Have you tested it on macOS? I am afraid that this command may not have all the necessary options on this system.

Reply:
There is no built-in parallel command on MacOS. This script is mainly for CI and having it run on MacOS is a nice-to-have (though useful). 

However I have taken care of it  and tested on macos (In the scrpits I made sure to install GNU Parallel from brew). do not use MacOS with Airflow - but if you do and you could double-check if this command works for MacOS, I would really appreciate.

```
function parallel::make_sure_gnu_parallel_is_installed() {
    start_end::group_start "Making sure GNU Parallels is installed"
    echo
    echo "Making sure you have GNU parallel installed"
    echo
    echo "You might need to provide root password if you do not have it"
    echo
    (command -v parallel || apt install parallel || sudo apt install parallel || brew install parallel) >/dev/null
    start_end::group_end "Making sure GNU Parallels is installed"
}
```

Reply:
Ok. I decided to take My Mac  to a spin and it was all surprisingly easy and straightforward. When  I gave my 64GB Mac - 40GB Memory + 8 CPUs it took 581 second (10 minutes) to run full suite of tests (!)

Parallel was not a problem, but I had to remove old commands where I checked memory and disk on host (which is not needed). I remove those and the script works flawlessly on Mac as well.